File: frontend/apps/marveline/src/types/product_variant.ts
// ProductVariant types — aligned with backend ProductVariantResponse (multi-dimensions)

export type ProductColor =
  | 'blanc'
  | 'ivoire'
  | 'bordeaux'
  | 'noir'
  | 'rouge'
  | 'vert_amande'
  | 'vert_sapin'
  | 'taupe'

export const PRODUCT_COLORS: { value: ProductColor; label: string }[] = [
  { value: 'blanc', label: 'Blanc' },
  { value: 'ivoire', label: 'Ivoire' },
  { value: 'bordeaux', label: 'Bordeaux' },
  { value: 'noir', label: 'Noir' },
  { value: 'rouge', label: 'Rouge' },
  { value: 'vert_amande', label: 'Vert amande' },
  { value: 'vert_sapin', label: 'Vert sapin' },
  { value: 'taupe', label: 'Taupe' },
]

export type ProductGamme =
  | 'classique'
  | 'elegance'
  | 'open_up'
  | 'prestige'
  | 'vintage'
  | 'bois'

export const PRODUCT_GAMMES: { value: ProductGamme; label: string }[] = [
  { value: 'classique', label: 'Classique' },
  { value: 'elegance', label: 'Élégance' },
  { value: 'open_up', label: "Open'Up" },
  { value: 'prestige', label: 'Prestige' },
  { value: 'vintage', label: 'Vintage' },
  { value: 'bois', label: 'Bois' },
]

export interface ProductVariant {
  id: number
  tenant_id: number
  product_id: number
  color?: string        // nullable — dimension optionnelle
  size?: string         // nullable — taille/format (ex: 21cm, 240cm)
  gamme?: string        // nullable — gamme/finition (ex: classique, elegance)
  label: string         // requis — clé d'unicité par produit, affiché partout
  price_per_day?: number // override prix parent en centimes (null = hérite parent)
  image_url?: string | null // override image (null = hérite parent)
  sku: string
  stock_quantity: number
  available_quantity: number
  weight_grams?: number | null
  volume_cm3?: number | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface ProductVariantCreate {
  color?: ProductColor  // optionnel
  size?: string         // optionnel
  gamme?: string        // optionnel
  label: string         // requis
  price_per_day?: number
  image_url?: string | null
  sku: string
  stock_quantity?: number
  available_quantity?: number
  weight_grams?: number | null
  volume_cm3?: number | null
}

export interface ProductVariantUpdate {
  color?: ProductColor
  size?: string
  gamme?: string
  label?: string
  price_per_day?: number
  image_url?: string | null
  stock_quantity?: number
  available_quantity?: number
  weight_grams?: number | null
  volume_cm3?: number | null
  is_active?: boolean
}

File: app/schemas/product_variant.py
"""Schémas Pydantic pour ProductVariant."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.constants.business import ProductColor


class ProductVariantCreate(BaseModel):
    """Données pour créer une variante produit multi-dimensions."""

    color: Optional[ProductColor] = Field(
        default=None,
        description="Couleur optionnelle (ex: blanc, ivoire)",
    )
    size: Optional[str] = Field(
        default=None,
        max_length=50,
        description="Taille/format optionnel (ex: 21cm, 240cm)",
    )
    gamme: Optional[str] = Field(
        default=None,
        max_length=50,
        description="Gamme/finition optionnelle (ex: classique, elegance)",
    )
    label: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Label affiché — clé d'unicité par produit",
    )
    price_per_day: Optional[int] = Field(
        default=None,
        ge=0,
        description="Override prix parent en centimes (None = hérite parent)",
    )
    sku: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="SKU unique de la variante (ex: NAP-RND-240-IVO)",
    )
    stock_quantity: int = Field(default=0, ge=0, description="Stock total")
    available_quantity: int = Field(
        default=0, ge=0, description="Quantité disponible"
    )
    deposit_amount: int = Field(
        default=0, ge=0, description="Montant caution par unité en centimes"
    )
    image_url: Optional[str] = Field(
        default=None, max_length=500, description="Override image URL (None = hérite parent)"
    )
    weight_grams: Optional[int] = Field(
        default=None, ge=0, description="Override poids en grammes (None = hérite parent)"
    )
    volume_cm3: Optional[int] = Field(
        default=None, ge=0, description="Override volume en cm³ (None = hérite parent)"
    )

    model_config = ConfigDict(use_enum_values=True)

    @model_validator(mode="after")
    def at_least_one_dimension(self) -> "ProductVariantCreate":
        """Au moins une dimension parmi color, size, gamme doit être renseignée."""
        if not any([self.color, self.size, self.gamme]):
            raise ValueError(
                "Au moins une dimension requise parmi : color, size, gamme"
            )
        return self


class ProductVariantUpdate(BaseModel):
    """Données pour mettre à jour une variante (PATCH partiel)."""

    color: Optional[ProductColor] = None
    size: Optional[str] = Field(default=None, max_length=50)
    gamme: Optional[str] = Field(default=None, max_length=50)
    label: Optional[str] = Field(default=None, min_length=1, max_length=100)
    price_per_day: Optional[int] = Field(default=None, ge=0)
    stock_quantity: Optional[int] = Field(default=None, ge=0)
    available_quantity: Optional[int] = Field(default=None, ge=0)
    deposit_amount: Optional[int] = Field(default=None, ge=0)
    image_url: Optional[str] = Field(default=None, max_length=500)
    weight_grams: Optional[int] = Field(default=None, ge=0)
    volume_cm3: Optional[int] = Field(default=None, ge=0)
    is_active: Optional[bool] = None

    model_config = ConfigDict(use_enum_values=True)


class ProductVariantNested(BaseModel):
    """Représentation minimale d'une variante pour inclusion dans les réponses nested."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    label: str
    color: Optional[str] = None
    size: Optional[str] = None
    gamme: Optional[str] = None
    sku: str
    price_per_day: Optional[int] = None
    image_url: Optional[str] = None
    weight_grams: Optional[int] = None
    volume_cm3: Optional[int] = None


class ProductVariantResponse(BaseModel):
    """Réponse API pour une variante produit."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: int
    product_id: int
    color: Optional[str]
    size: Optional[str]
    gamme: Optional[str]
    label: str
    price_per_day: Optional[int]
    sku: str
    stock_quantity: int
    available_quantity: int
    deposit_amount: int
    image_url: Optional[str] = None
    weight_grams: Optional[int] = None
    volume_cm3: Optional[int] = None
    is_active: bool
    created_at: datetime
    updated_at: datetime
